airflow/models/taskmixin.py

```diff
@@ -15,61 +15,78 @@
 # specific language governing permissions and limitations
 # under the License.
 
+import warnings
 from abc import abstractmethod
 from typing import Sequence, Union
 
+DependencyMixinOrList = Union["DependencyMixin", Sequence["DependencyMixin"]]
 
-class TaskMixin:
-    """
-    Mixing implementing common chain methods like >> and <<.
 
-    In the following functions we use:
-    Task = Union[BaseOperator, XComArg]
-    No type annotations due to cyclic imports.
-    """
+class DependencyMixin:
+    """Mixing implementing common dependency setting methods methods like >> and <<."""
 
     @property
-    def roots(self):
-        """Should return list of root operator List[BaseOperator]"""
+    def roots(self) -> Sequence["DependencyMixin"]:
+        """
+        List of root nodes -- ones with no upstream dependencies.
+
+        a.k.a. the "start" of this sub-graph
+        """
         raise NotImplementedError()
 
     @property
-    def leaves(self):
-        """Should return list of leaf operator List[BaseOperator]"""
+    def leaves(self) -> Sequence["DependencyMixin"]:
+        """
+        List of leaf nodes -- ones with only upstream dependencies.
+
+        a.k.a. the "end" of this sub-graph
+        """
         raise NotImplementedError()
 
     @abstractmethod
-    def set_upstream(self, other: Union["TaskMixin", Sequence["TaskMixin"]]):
+    def set_upstream(self, other: DependencyMixinOrList):
```

Comment:
I'm not 100% sold on this name, but it felt very verbose the number of places we had `Union["TaskMixin", Sequence["TaskMixin"]]` used.

Is this actually any easier to understand, or is the indirection just hiding this and making it harder to grok?

Reply:
one vote for verbose in this case
eventually it would become `"TaskMixin" | Sequence["TaskMixin"]` which maybe ain’t as bad?

Reply:
Mmmmm, that's 3.9 only isnt it, so a way off, but I think I agree on this alias not helping.

Reply:
yeah long way off, true

Reply:
https://mypy.readthedocs.io/en/stable/runtime_troubles.html#future-annotations-import-pep-563 -- form 3.7 and from-future-annotations we can have it I think! Which is just weeks away.